doc/design/design.md

```diff
@@ -0,0 +1,245 @@
+  * Everything validated in the client (validate again to prevent attacks)
+  * Validation of the whole input: Here we can validate more complex constraints that for example involve multiple
+    fields
+
+For the validation done by the client we define the constraints declarative in the extension:
+
+```ts
+param = {
+  id: "",
+  name: "",
+  type: "string",
+  regex: /\d+/ // <-- Validation via regex
+}
+```
+
+For the more complex validation we define it via a JS callback function. That gives the extension definition the maximum
+flexibility.
+The validation code is executed in the JS VM in the backend.
+
+```ts
+extension = {
+  //...
+  validate(parameterValues) {
+    if (parameterValues["source"] === "s3" && !isValidBucketName(parameterValues["bucket"])) {
+      return "Invalid bucket name.";
+    }
+  }
+}
+```
+
+We decided against a callback for the client side validation. The main reason was that it would require to load code
+into the client at runtime. That would require to run code `eval` on code retrieved from a request. Even so the security
+risk of that seems acceptable, we decided against it since it would look suspicious.
+
+## Deployment of the Extension Definitions
+
+During the build of each virtual-schema project the extension definitions (written in TypeScript) are compiled to a
+single JavaScript file:
+
+```plantuml
+file postgres-extension-def.ts{
+}
+
+file postgres-extension-def.js{ 
+}
+
+"postgres-extension-def.ts" -> "postgres-extension-def.js": compiled during build
+```
+
+The built installers are attached to the release:
+
+```
+Release assets:
+postgresql-vs-2.0.2.jar
+postgresql-vs-2.0.2.jar.sha256
+postgres-extension-def-2.0.2.js
+```
+
+A crawler collects the JARs and extension definitions and copies them to BucketFS:
+The crawler runs at scheduled intervals.
+The crawler is at the moment not part of this project.
+
+```plantuml
+storage "prostgres-virtual-schema release 2.0.2"{
+file installerInRepo as "postgres-extension-def-2.0.2.js"
+file jarV2InRepo as "postgresql-vs-2.0.2.jar"
+}
+
+storage "prostgres-virtual-schema release 1.0.0"{
+file "postgres-extension-def-1.0.0.js"
+file jarV1InRepo as "postgresql-vs-1.0.0.jar"
+}
+
+node "Exasol SaaS backend"{
+storage Installers{
+file installerInBucketFS as "postgres-extension-def-2.0.2.js"
+}
+}
+
+node "Exasol cluster (per customer)"{
+storage BucketFS{
+file jarV2InBucketFs as "postgresql-vs-2.0.2.jar"
+file jarV1InBucketFS as "postgresql-vs-1.0.0.jar"
+}
+}
+
+installerInRepo --> installerInBucketFS
+jarV2InRepo --> jarV2InBucketFs
+jarV1InRepo --> jarV1InBucketFS
+```
+
+Note that only the latest version of the extension definition is deployed.
+That does not mean that it's impossible to install older versions of an extension. Just the extension definition (the
+plugin for the extensions store) is always the latest version.
+That means that the extension definition must be able to uninstall and update all old versions of its extension. That
```

Comment:
So, if I understand this correctly, it needs to be able to do this, but it doesn't do it automatically, when a new version is available.

Reply:
for now yes. In the future maybe we add a feature for automatic updated but then we can make it optional.